File: app/src/main/java/com/storehop/app/data/entity/ItemStoreXref.kt
package com.storehop.app.data.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index

@Entity(
    tableName = "item_store_xref",
    primaryKeys = ["itemId", "storeId"],
    foreignKeys = [
        ForeignKey(
            entity = Item::class,
            parentColumns = ["id"],
            childColumns = ["itemId"],
            onDelete = ForeignKey.CASCADE,
        ),
        ForeignKey(
            entity = Store::class,
            parentColumns = ["id"],
            childColumns = ["storeId"],
            onDelete = ForeignKey.CASCADE,
        ),
    ],
    indices = [
        Index("storeId"),
        Index("userId"),
        Index("householdId"),
        Index("deletedAt"),
    ],
)
data class ItemStoreXref(
    val itemId: String,
    val storeId: String,
    val userId: String,
    val createdAt: Long,
    val updatedAt: Long,
    val deletedAt: Long?,
    @ColumnInfo(defaultValue = "1") val pendingSync: Boolean = true,
    /**
     * Per-store need state. True means "still need to buy this here." False
     * means "I bought this here."
     *
     * The *storage* is per-(item, store), but the *check-off cascade* is
     * global: one shopping trip satisfies the need everywhere. Buying milk at
     * Lidl cascades `isNeeded = 0` to milk's Aldi xref too
     * ([ItemStoreXrefDao.markPurchasedAcrossAllStores]); un-buying it at Lidl
     * cascades `isNeeded = 1` back across every tagged store
     * ([ItemStoreXrefDao.markNeededAcrossAllStores]). The per-store column
     * still lets the strike-through window and staple visibility differ per
     * screen, but "needed" is effectively one state shared across the item's
     * stores.
     */
    @ColumnInfo(defaultValue = "1") val isNeeded: Boolean = true,
    /**
     * When this specific (item, store) pair was last marked purchased.
     * Powers the within-session strike-through window on the Shop-at-Store
     * screen so the user can undo a mis-tap before leaving the store.
     */
    val lastPurchasedAt: Long? = null,
    /** v0.7.0 multi-user household access scope. See [Item.householdId]. */
    @ColumnInfo(defaultValue = "''") val householdId: String = "",
)

File: ios/Storehop/Data/Models/ItemStoreXref.swift
import Foundation
import GRDB

/// One row per (item, store) pair that the user has tagged together. Owns the
/// per-store shopping state since migration v4→v5: `isNeeded` says whether the
/// item is still needed at this specific store, and `lastPurchasedAt` records
/// when it was last marked purchased there. Marking purchased at one store
/// cascades `isNeeded = false` across every live xref for the item (the
/// v0.5.1 cross-store cascade), but `lastPurchasedAt` is set only on the
/// store where the user actually shopped.
struct ItemStoreXref: Codable, FetchableRecord, MutablePersistableRecord, Hashable, Sendable {
    static let databaseTableName = "item_store_xref"

    var itemId: String
    var storeId: String
    var userId: String
    var createdAt: Int64
    var updatedAt: Int64
    var deletedAt: Int64?
    var pendingSync: Bool
    var isNeeded: Bool
    var lastPurchasedAt: Int64?
    /// v0.7.0 multi-user access scope. See `Item.householdId` for the full
    /// rationale; single-member households have `householdId == userId`.
    var householdId: String = ""

    enum Columns {
        static let itemId          = Column(CodingKeys.itemId)
        static let storeId         = Column(CodingKeys.storeId)
        static let userId          = Column(CodingKeys.userId)
        static let createdAt       = Column(CodingKeys.createdAt)
        static let updatedAt       = Column(CodingKeys.updatedAt)
        static let deletedAt       = Column(CodingKeys.deletedAt)
        static let pendingSync     = Column(CodingKeys.pendingSync)
        static let isNeeded        = Column(CodingKeys.isNeeded)
        static let lastPurchasedAt = Column(CodingKeys.lastPurchasedAt)
        static let householdId     = Column(CodingKeys.householdId)
    }
}
